ipld/unixfs/io/directory.go

```diff
@@ -472,14 +658,32 @@ func (d *HAMTDirectory) needsToSwitchToBasicDir(ctx context.Context, name string
 		operationSizeChange += linksize.LinkSizeFunction(name, nodeToAdd.Cid())
 	}
 
-	if d.sizeChange+operationSizeChange >= 0 {
-		// We won't have reduced the HAMT net size.
-		return false, nil
+	// We must switch if size and maxlinks are below threshold
+	canSwitchSize := false
+	// Directory size reduced, perhaps below limit.
+	if d.sizeChange+operationSizeChange < 0 {
+		canSwitchSize, err = d.sizeBelowThreshold(ctx, operationSizeChange)
+		if err != nil {
+			return false, err
+		}
 	}
 
-	// We have reduced the directory size, check if went below the
-	// HAMTShardingSize threshold to trigger a switch.
-	return d.sizeBelowThreshold(ctx, operationSizeChange)
+	canSwitchMaxLinks := true
+	if d.maxLinks > 0 {
+		total := d.totalLinks
+		if nodeToAdd != nil && entryToRemove == nil {
+			total++
+		} else if nodeToAdd == nil && entryToRemove != nil {
+			total--
+		}
```

Comment:
Are the extra checks worth it?
```suggestion
		if nodeToAdd != nil {
			total++
		}
		if entryToRemove != nil {
			total--
		}
```